Added doc for Ztunnel inpod mode + updated ztunnel L4 guide for 1.21 function (#14611)
istio/istio.io: content/en/docs/ops/ambient/usage/inpod/index.md

---
title: ZTunnel Inpod traffic redirection 
description: User guide for details on the ZTunnel inpod traffic redirection function.
weight: 2
owner: istio/wg-networking-maintainers
test: no
---

## Introduction {#introsection}

This guide is a supplement to the [Istio Ambient L4 Networking guide](/docs/ops/ambient/usage/ztunnel) and provides additional detail on the traffic redirection mechanism between endpoints and ZTunnel proxies. It is recommended to read that guide prior to reading this one.

## Ambient traffic redirection

> In the context of Istio Ambient, the term traffic redirection refers to data plane functionality that intercepts traffic sent to and from endpoints that ambient enabled and the proxies that handle the core data path in the ambient mesb. Sometimes the term "traffic capture" is also used equivalently, in this guide we will stick with the term "traffic redirection". 

Reviewer: ```suggestion
In the context of ambient mode, _traffic redirection_ refers to data plane functionality that intercepts traffic sent to and from ambient-enabled workloads, routing it through the node proxies that handle the core data path. Sometimes the term _traffic capture_ is also used.
```

> In the context of Istio Ambient, the term traffic redirection refers to data plane functionality that intercepts traffic sent to and from endpoints that ambient enabled and the proxies that handle the core data path in the ambient mesb. Sometimes the term "traffic capture" is also used equivalently, in this guide we will stick with the term "traffic redirection". 

Reviewer: I took the best of this comment...

> In the context of Istio Ambient, the term traffic redirection refers to data plane functionality that intercepts traffic sent to and from endpoints that ambient enabled and the proxies that handle the core data path in the ambient mesb. Sometimes the term "traffic capture" is also used equivalently, in this guide we will stick with the term "traffic redirection". 

Reviewer: but not this one, because you missed "mesb" 🤣

{{< /tip >}}

### Inpod traffic redirection model

The core design principle behind the new inpod traffic redirection function is that the ztunnel proxy has the ability to perform all its data path required functionality directly inside the linux network namespace of an ambient endpoint/ pod. This is achieved via a combination of new functionality added to the `istio-cni` node agent and the ztunnel proxy components from the 1.21.0 release onwards.  A key benefit of this model is that it enables Istio Ambient to work alongside any Kubernetes CNI plugin transparently and without impacting kubernetes networking features. 

The following figure illustrates the sequence of steps that set this up when a new ambient pod is created or converted to ambient mode. 

{{< image width="100%"
link="./pod-added-to-ambient.svg"
alt="pod added to the ambient mesh flow"
>}}

> The `istio-cni` node agent is informed of pod lifecycle events such as creation and deletion and also watches the underlying Kubernetes api server for events such as the ambient label being attached to a pod or its namespace. The `istio-cni` is a CNI plugin that is secondary to the primary CNI plugin within that Kubernetes cluster. Its purpose is only to setup traffic redirection and istio specific functions while letting the primary Kubernetes CNI plugin handle core L3 networking setup that is needed independent of istio. Once the `istio-cni` node agent is informed of sych an event that requires it to be added to the ambient mesh, the following sequence of operations is performed.

Reviewer 2: ```suggestion
The `istio-cni` node agent is informed of pod lifecycle events such as creation and deletion, and also watches the underlying Kubernetes API server for events such as the ambient label being added to a pod or namespace. 

The `istio-cni` node agent additionally installs a chained CNI plugin that is executed by the container runtime after the primary CNI plugin within that Kubernetes cluster executes. Its only purpose is to notify the `istio-cni` node agent when a new pod is created by the container runtime in a namespace that is already enrolled in ambient mode, and propagate the new pod context to the `istio-cni` node agent. 

Once the `istio-cni` node agent is notified that a pod needs to be added to the ambient mesh (either from the CNI plugin, if the pod is brand-new, or from the Kubernetes API server, if the pod is already running but needs to be added), the following sequence of operations is performed:
```

> i

Reviewer 2: ```suggestion

```

network namespace to create listening sockets in another network namespace, assuming the target network namespace is known
at creation time.
- The node-local ztunnel internally spins up a new proxy instance and listen port set, dedicated to the newly-added pod.
- Once the in-Pod redirect rules are in place and the ztunnel has established the listen ports, the pod is added in the
mesh and traffic begins flowing thru the node-local ztunnel, as before.
 
Once the pod is successfully added to the ambient mesh, traffic to and from pods in the mesh will be fully encrypted with mTLS by default, as always with Istio.

Traffic will now enter and leave the pod network namespace as encrypted traffic - it will look like every pod in the ambient mesh has the ability to enforce mesh policy and securely encrypt traffic, even though the user application running in the pod
has no awareness of either.

Here’s a diagram to illustrate how encrypted traffic flows between pods in the ambient mesh in the new model:

{{< image width="100%"
    link="./traffic-flows-between-pods-in-ambient.svg"
    alt="HBONE traffic flows between pods in the ambient mesh"
    >}}

> ### Observing the inpod traffic redirection in an ambient mesh

Reviewer 2: ```suggestion
### Observing and debugging the inpod traffic redirection in an ambient mesh
```

> If Istio Ambient traffic is not working correctly, there are some quick checks that can be made to help narrow down the problem.  In order to observe inpod traffic redirection in action, first follow the steps described in the [ztunnel L4 networking guide](/docs/ops/ambient/usage/ztunnel) including deployment of istio ambient mesh on a Kubernetes Kind cluster and the deployment of the `httpbin` and `sleep` deployments in the namespaced tagged for ambient operation as described in that guide. Once it is verified that the application is successfully running in the ambient mesh, use the following steps to observe the inpod traffic redirection.

Reviewer 3: ```suggestion
If Istio Ambient traffic is not working correctly, some quick checks can be made to help narrow down the problem.  To observe inpod traffic redirection in action, first follow the steps described in the [ztunnel L4 networking guide](/docs/ops/ambient/usage/ztunnel), including deployment of Istio ambient mode on a Kubernetes Kind cluster and the deployment of the `httpbin` and `sleep` deployments in the namespace tagged for ambient operation as described in that guide. Once it's verified that the application is successfully running in the ambient mesh, you can use the following steps to observe the inpod traffic redirection.
```

> In order to confirm that the mesh is using Inpod redirection, once Ambient has been enabled in an application pod, check the ztunnel proxy logs. As shown in the example below, the ztunnel logs related to inpod indicate that inpod mode is enabled and ztunnel proxy has received the netns information about an ambient application pod and has started proxying for it. 

Reviewer 3: ```suggestion

1.  Check the ztunnel proxy logs

When an application pod is part of the ambient mode you can check the ztunnel proxy logs to confirm the mesh uses inpod redirection. As shown in the example below, the ztunnel logs related to inpod indicate that inpod mode is enabled and ztunnel proxy has received the network namespace (netns) information about an ambient application pod and has started proxying for it. 
```

{{< text bash >}}
$ kubectl logs ds/ztunnel -n istio-system  | grep inpod
Found 3 pods, using pod/ztunnel-hl94n
inpod_enabled: true
inpod_uds: /var/run/ztunnel/ztunnel.sock
inpod_port_reuse: true
inpod_mark: 1337
2024-02-21T22:01:49.916037Z  INFO ztunnel::inpod::workloadmanager: handling new stream
2024-02-21T22:01:49.919944Z  INFO ztunnel::inpod::statemanager: pod WorkloadUid("1e054806-e667-4109-a5af-08b3e6ba0c42") received netns, starting proxy
2024-02-21T22:01:49.925997Z  INFO ztunnel::inpod::statemanager: pod received snapshot sent
2024-02-21T22:03:49.074281Z  INFO ztunnel::inpod::statemanager: pod delete request, draining proxy
2024-02-21T22:04:58.446444Z  INFO ztunnel::inpod::statemanager: pod WorkloadUid("1e054806-e667-4109-a5af-08b3e6ba0c42") received netns, starting proxy
{{< /text >}}

> You should also be able to confirm that sockets are open and in listening state on ports 15001, 15006 and 15008 within an application's network namespace as follows.

Reviewer 3: ```suggestion
2. Confirm the state of sockets

Please follow the steps below to confirm that the sockets on ports 15001, 15006, and 15008 are open and in the listening state.
```

{{< text bash >}}
$ kubectl debug $(kubectl get pod -l app=sleep -n ambient-demo -o jsonpath='{.items[0].metadata.name}') -it -n ambient-demo  --image nicolaka/netshoot  -- ss -ntlp
Defaulting debug container name to debugger-nhd4d.
State  Recv-Q Send-Q Local Address:Port  Peer Address:PortProcess
LISTEN 0      128        127.0.0.1:15080      0.0.0.0:*
LISTEN 0      128                *:15006            *:*
LISTEN 0      128                *:15001            *:*
LISTEN 0      128                *:15008            *:*
{{< /text >}}

> In order to view the iptables rules setup inside one of the application pods, execute the command shown in the following example. The command output shows that additional Istio specific chains are added into the NAT and Mangle tables in netfilter/ iptables within the application pod's network namespace and traffic that is coming into the pod on TCP port 15008 or 15006 is redirected into the ztunnel proxy for ingress Ambient processing whereas traffic leaving the pod is redirected to ztunnel's port 15001 for egress processing prior to being sent out by ztunnel using HBONE encapsulation.

Reviewer 3: ```suggestion
3. Check the iptables rules setup

To view the iptables rules setup inside one of the application pods, execute the command shown in the following example. 

The command output shows that additional Istio-specific chains are added to the NAT and Mangle tables in netfilter/iptables within the application pod's network namespace. Traffic coming into the pod on TCP port 15008 or 15006 is redirected into the ztunnel proxy for ingress Ambient processing. The traffic leaving the pod is redirected to ztunnel's port 15001 for egress processing before being sent out by ztunnel using HBONE encapsulation.
```
